packages/core/src/common/registry/resolve-id-registry.ts

```diff
@@ -45,7 +47,11 @@ export default class ResolveIdRegistry<TType, TToken extends { [key: string]: un
                 }
             }
 
-            results.push(result);
+            const keys = Object.keys(resolverId).length;
+
+            if (keys === result.matches || result.default === true) {
+                results.push(result);
+            }
```

Comment:
I undertand this change limits matching results to those with the exact same key names or the default method, which seems too strict. What do you think?

Reply:
I was thinking the same when I was doing that, does it make sense to just add a check for `gateway` in the keys list?

Reply:
Are the e2e tests in `checkout-js` good with this PR?

Reply:
I haven't tried running them as of now. But I suppose should be ok.

Reply:
The PR description doesn't seem to match with what the actual problem described in #1917? The query is always going to have all the keys, e.g.: 
```json
{"id":"credit_card", "gateway":"barclaycard", "type":"PAYMENT_TYPE_HOSTED"}
```
instead of 
```json
{"id":"credit_card"}
```
The resolve ID, however, may not contain all the keys. So it could be `{"id":"credit_card", "gateway":"bluesnapdirect"}` or `{"type":"PAYMENT_TYPE_HOSTED"}` or `{"id":"credit_card"}`. 

Based on that problem definition, I think what we need is change the `for` loop in `resolveToken` to something like this?
```ts
for (const [key, value] of Object.entries(resolverId)) {
    if (key in query && query[key] !== value) {
        result.matches = 0;
        break;
    }

    if (key in query && query[key] === value) {
        result.matches++;
    }

    // ...
}
```

With this, if a strategy is specifically registered for `{"id":"credit_card", "gateway":"bluesnapdirect"}`, it won't match with `{"id":"credit_card", "gateway":"barclaycard", "type":"PAYMENT_TYPE_HOSTED"}` as the `gateway` field is incompatible. But if the strategy is registered for `{"id":"credit_card"}`, it will still match as there is no incompatible field.